### PR TITLE
Add `datetime::is_leap_year`

### PR DESCRIPTION
**Pull request by isVoid**
Part 1 of #8677 

This PR adds `datetime::is_leap_year`. The function returns `true` for datetime column rows with leap year; `false` for rows with non leap years, and `null` for null rows.

### CODE REVIEW

File: cpp/include/cudf/detail/datetime.hpp [hunk 1]
```diff
@@ -124,6 +124,16 @@ std::unique_ptr<cudf::column> add_calendrical_months(
...
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
+
+/**
+ * @copydoc cudf::is_leap_year(cudf::column_view const&, rmm::mr::device_memory_resource *)
+ *
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ */
+std::unique_ptr<cudf::column> is_leap_year(
+  cudf::column_view const& column,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
```
Comment (karthikeyann) [inline cpp/include/cudf/detail/datetime.hpp:135]: `stream` argument is missing.
```suggestion
  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
```

File: cpp/src/datetime/datetime_ops.cu [hunk 2]
```diff
@@ -357,6 +358,30 @@ std::unique_ptr<column> day_of_year(column_view const& column,
...
+std::unique_ptr<column> is_leap_year(column_view const& column,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(is_timestamp(column.type()), "Column type should be timestamp");
+
+  auto years      = detail::extract_year(column, stream);
+  auto years_view = years->view();
+
+  auto mask   = copy_bitmask(column.null_mask(), column.offset(), column.offset() + column.size());
```
Comment (karthikeyann) [inline cpp/src/datetime/datetime_ops.cu:371]: use `stream` and `mr` for `copy_bitmask`
```suggestion
  auto mask   = detail::copy_bitmask(column, stream, mr);
```

File: cpp/include/cudf/datetime.hpp [hunk 3]
```diff
@@ -189,6 +189,23 @@ std::unique_ptr<cudf::column> add_calendrical_months(
   cudf::column_view const& timestamps,
   cudf::column_view const& months,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
+
+/**
+ * @brief  Check if the year of the given date is a leap year
+ *
+ * `output[i] == true` if year of `column[i]` is a leap year
+ * `output[i] == false` if year of `column[i]` is not a leap year
+ * `output[i] is null` if year of `column[i]` is null
```
Comment (karthikeyann) [inline cpp/include/cudf/datetime.hpp:198]:
```suggestion
 * `output[i] is null` if `column[i]` is null
```

File: cpp/src/datetime/datetime_ops.cu [hunk 4]
```diff
@@ -19,6 +19,7 @@
 #include <cudf/column/column_factories.hpp>
 #include <cudf/column/column_view.hpp>
 #include <cudf/datetime.hpp>
+#include <cudf/detail/datetime.hpp>
 #include <cudf/detail/null_mask.hpp>
 #include <cudf/detail/nvtx/ranges.hpp>
 #include <cudf/null_mask.hpp>
```
Comment (karthikeyann) [inline cpp/src/datetime/datetime_ops.cu:25]:
```suggestion
```

**Review by karthikeyann — CHANGES_REQUESTED** [review]

File: cpp/src/datetime/datetime_ops.cu [hunk 5]
```diff
@@ -357,6 +358,30 @@ std::unique_ptr<column> day_of_year(column_view const& column,
...
 }
+
+std::unique_ptr<column> is_leap_year(column_view const& column,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(is_timestamp(column.type()), "Column type should be timestamp");
+
+  auto years      = detail::extract_year(column, stream);
+  auto years_view = years->view();
```
Comment (jrhemstad) [inline cpp/src/datetime/datetime_ops.cu:369]: There's no need to materialize the column of years. You can extract the year on the fly inside your transform kernel below.

**Comment by isVoid** [comment]
The implementation has changed to largely reusing existing `apply_datetime_op` scaffolding.

**Review by harrism — APPROVED** [review]
Looks great.

**Review by karthikeyann — APPROVED** [review]

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**